packages/x-scheduler/src/joy/internals/components/time-grid/TimeGrid.tsx

```diff
@@ -0,0 +1,161 @@
+'use client';
+import * as React from 'react';
+import clsx from 'clsx';
+import { SchedulerValidDate } from '@mui/x-scheduler/primitives/utils/adapter/types';
+import { getAdapter } from '../../../../primitives/utils/adapter/getAdapter';
+import { TimeGrid as TimeGridPrimitive } from '../../../../primitives/time-grid';
+import { TimeGridProps } from './TimeGrid.types';
+import { CalendarEvent } from '../../../models/events';
+import { Event } from '../../../event/Event';
+import { isWeekend } from '../../../utils/date-utils';
+import { useTranslations } from '../../../utils/TranslationsContext';
+import './TimeGrid.css';
+
+const adapter = getAdapter();
+
+export const TimeGrid = React.forwardRef(function TimeGrid(
+  props: TimeGridProps,
+  forwardedRef: React.ForwardedRef<HTMLDivElement>,
+) {
+  const { events, days, className, onDayHeaderClick, ...other } = props;
+
+  const translations = useTranslations();
+  const today = adapter.date('2025-05-26');
+  const bodyRef = React.useRef<HTMLDivElement>(null);
+  const headerWrapperRef = React.useRef<HTMLDivElement>(null);
+
+  const eventsByDay = React.useMemo(() => {
+    const map = new Map();
+    for (const event of events) {
+      const dayKey = adapter.format(event.start, 'keyboardDate');
+      if (!map.has(dayKey)) {
+        map.set(dayKey, []);
+      }
+      map.get(dayKey).push(event);
+    }
+    return map;
+  }, [events]);
+
+  React.useLayoutEffect(() => {
+    const body = bodyRef.current;
+    const header = headerWrapperRef.current;
+    if (!body || !header) {
+      return;
+    }
+    const hasScroll = body.scrollHeight > body.clientHeight;
+    header.style.setProperty('--has-scroll', hasScroll ? '1' : '0');
+  }, [events]);
+
+  const lastIsWeekend = isWeekend(adapter, days[days.length - 1]);
+
+  const handleHeaderClick = React.useCallback(
+    (day: SchedulerValidDate) => (event: React.MouseEvent) => {
+      onDayHeaderClick?.(day, event);
+    },
+    [onDayHeaderClick],
+  );
+
+  const renderHeaderContent = (day: SchedulerValidDate) => (
+    <span className="TimeGridHeaderContent">
+      {/* TODO: Add the 3 letter week day format to the adapter */}
+      <span className="TimeGridHeaderDayName">{adapter.formatByString(day, 'ccc')}</span>
+      <span className="TimeGridHeaderDayNumber">{adapter.format(day, 'dayOfMonth')}</span>
+    </span>
+  );
+
+  return (
+    <div ref={forwardedRef} className={clsx('TimeGridContainer', 'joy', className)} {...other}>
+      <TimeGridPrimitive.Root className="TimeGridRoot">
+        <div ref={headerWrapperRef} className="TimeGridHeader">
+          <div className="TimeGridGridRow TimeGridHeaderRow" role="row">
```

Comment:
Nitpick: Wouldn't this make sense as a primitive @flaviendelangle? 🤔

Reply:
Added for grooming #18332